Automattic/newspack-blocks
# feat(ras-acc): add newsletter signup option to checkout button and donation blocks

### All Submissions:






### Changes proposed in this Pull Request:



Closes https://app.asana.com/0/1206810965479792/1205992141743582

This PR adds a newsletter subscription option to the donate and checkout button blocks. When enabled, the selected list will be presented to readers when checking out via the blocks. There is also an additional option to hide the subscription form and automatically subscribe the reader:

Block editor:
![Screenshot 2024-04-15 at 16 18 00](https://github.com/Automattic/newspack-blocks/assets/17905991/aa2a9f34-71e1-49a5-a774-41d5ce72380d)

Frontend:
![Screenshot 2024-04-12 at 18 03 44](https://github.com/Automattic/newspack-blocks/assets/17905991/1c9ecd29-9328-45a5-b363-d3802c7d53fa)

### How to test the changes in this Pull Request:

Before testing, make sure you also have `newspack-plugin` checked out to the `epic/ras-acc` branch.

First, we will test the blocks in the editor:
1. Go to any page or post and add a checkout button block.
2. In the block settings look for a new newsletter subscription section (pictured above)
3. Confirm each of the toggles can be toggled and updated to persist state:
4. Repeat the above steps with the donate block

Next, we will test the frontend functionality:
1. Visit the page or post where you added the blocks in the above steps.
2. Select either one and go through the full checkout process.
3. Depending on the settings you selected when setting up the block, confirm a newsletter signup form either appears or doesn't appear:
    - Enable newsletter subscription: Enables newsletter subscription during the checkout process for the block. Will cause the subscription form to appear UNLESS the next option is set.
    - Hide newsletter selection and always subscribe: Hides the subscription form but automatically subscribes the customer to the selected list
    - Lists: The selected lists will appear on the newsletter subscription form. If the form is hidden via the hide option, the customer will automatically be subscribed to the selected lists.
4. Test out the checkout flow with various newsletter subscription setting configurations (disable subscriptions, and ensure the form doesn't appear. Enable the form but select the hide option and confirm the contact is added to the esp, etc.)
5. Repeat the above steps for the donation block
6. Ensure you have a contact that has already been subscribed in the ESP. Go through the checkout flow for either block and confirm the form doesn't appear regardless of newsletter subscription settings

Finally, we will test the blocks still function when newsletters is inactive
1. Deactivate the newsletters plugin
2. Repeat the above steps, but make sure the option to set newsletter subscription settings is not present in block editors AND no subscription form appears during the checkout flow on the frontend.

### Other information:
